### PR TITLE
Runs tests in blame mode

### PR DESCRIPTION
**Pull request by russcam**
This commit runs tests in blame mode with a blame-hang timeout of 5 minutes. This is to capture cases where test runs hang on CI.

### CODE REVIEW

**Comment by russcam** [comment]
jenkins run the tests

**Comment by russcam** [comment]
jenkins run the tests

**Review by gregkalapos — CHANGES_REQUESTED** [review]
I went through this and I agree that this can potentially be related to the activity id format and to https://github.com/elastic/apm-agent-dotnet/pull/896 and https://github.com/elastic/apm-agent-dotnet/pull/1022. But I'm still not 100% sure.

If the theory is true, then this should consistently fail - The bug where the `.End` events are not sent should basically always happen - at least when I reproduced it, it was always the case.

The `HttpCallWithW3CActivityFormat` test is now failing on Windows (both .NET Core and on Full Framework). I see you added a timeout there `mockPayloadSender.WaitForSpans(TimeSpan.FromMinutes(1));`. So this makes me think now instead of hanging, we got failing tests after 1 minute - so this proves that those tests are indeed not ok.

I tried to figure out why this happens, but locally that test is green for me on Windows. This original problem where we don't get the `End` event and the span is missing is related to the version of `System.Diagnostics.DiagnosticSource` and if it's pre `5.x`, I think we'll have the problem (see https://github.com/elastic/apm-agent-dotnet/pull/1022). Could it be that in CI we somehow end up with an older version? From what I see it should always be `5.x`, unless it's build for the startuphook - or isn't that the case? 🤔

File: test/Elastic.Apm.Tests/BackendCommTests/CentralConfig/CentralConfigFetcherTests.cs [hunk 1]
```diff
@@ -67,9 +67,9 @@ public void Should_Update_Logger_That_Is_ILogLevelSwitchable()
 			centralConfigFetcher.IsRunning.Should().BeTrue();
 			waitHandle.WaitOne();
 
-			// wait up to 20 seconds for the log level to change. Change can often be slower in CI
+			// wait up to 60 seconds for the log level to change. Change can often be slower in CI
 			var count = 0;
-			while (count < 20 && testLogger.LogLevelSwitch.Level == logLevel)
+			while (count < 60 && testLogger.LogLevelSwitch.Level == logLevel)
```
Comment (gregkalapos) [inline test/Elastic.Apm.Tests/BackendCommTests/CentralConfig/CentralConfigFetcherTests.cs:72]: 👍 Indeed we had a flaky test, likely related to this.

**Comment by russcam** [comment]
`HttpCallWithW3CActivityFormat` is also green for me locally. 

I've added an `XUnitLogger` APM logger implementation to capture the agent logs for the test and also emit the version of `System.Diagnostics.DiagnosticSource`; it's running with 5.0.0 when it fails. I'm wondering if there's some concurrent cross-test interference/pollution happening when this test runs that sometimes results in failure. [On one failing test run](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp/detail/PR-1193/20/tests), it looks like the diagnostic listener didn't get called for the `HttpClient` call in the test (it did for central configuration though):

```
HttpCallWithW3CActivityFormat: {SystemInfoHelper} No /proc/self/cgroup found - the agent will not report container id
HttpCallWithW3CActivityFormat: {HttpDiagnosticListener} Current runtime is detected as Full Framework. RuntimeInformation.FrameworkDescription: .NET Framework 4.8.4018.0
HttpCallWithW3CActivityFormat: {DiagnosticInitializer} Subscribed Elastic.Apm.DiagnosticListeners.HttpDiagnosticListenerFullFrameworkImpl to `System.Net.Http.Desktop' events source
HttpCallWithW3CActivityFormat: {AbstractConfigurationReader (MockConfigSnapshot)} No transaction sample rate provided. Defaulting to '1'
HttpCallWithW3CActivityFormat: {Transaction} New Transaction instance created: Transaction{Id: 3720f0183d694847, TraceId: e07304dcd1d4c845bbdc296fe72fa4ec, ParentId: null, Name: Test, Type: Test, Outcome: Unknown, IsSampled: True}. IsSampled (True) is based on the given sampler (Sampler{ constant: True }). Start time: 2021-02-23 04:17:13.9845550 UTC (as timestamp: 1614053833984555)
HttpCallWithW3CActivityFormat: {Tracer} Starting Transaction{Id: 3720f0183d694847, TraceId: e07304dcd1d4c845bbdc296fe72fa4ec, ParentId: null, Name: Test, Type: Test, Outcome: Unknown, IsSampled: True}
HttpCallWithW3CActivityFormat: {Transaction} Ended Transaction{Id: 3720f0183d694847, TraceId: e07304dcd1d4c845bbdc296fe72fa4ec, ParentId: null, Name: Test, Type: Test, Outcome: Unknown, IsSampled: True}. Start time: 2021-02-23 04:17:13.9845550 UTC (as timestamp: 1614053833984555), End time: 2021-02-23 04:17:13.9943170 UTC (as timestamp: 1614053833994317), Duration: 9.762ms
HttpCallWithW3CActivityFormat: {HttpDiagnosticListenerImplBase} Called with key: `System.Net.Http.Desktop.HttpRequestOut.Start'
HttpCallWithW3CActivityFormat: {AbstractConfigurationReader (MockConfigSnapshot)} Using default ServerUrl: http://localhost:8200/
HttpCallWithW3CActivityFormat: {HttpDiagnosticListenerImplBase} Request URL (http://localhost:8200/config/v1/agents?service.name=xunit_execution_desktop) is filtered out - exiting
```

**Comment by russcam** [comment]
I've kicked the tests off again.

Digging into `HttpHandlerDiagnosticListener`, The `"System.Net.Http.Desktop.HttpRequestOut.Ex.Stop"` key will be raised when the response content length is 0 and the request is async, which is true for the HttpCallWithW3CActivityFormat test

https://github.com/dotnet/runtime/blob/7565d60891e43415f5e81b59e50c52dba46ee0d7/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs#L543-L564

```c#
                        // In case reponse content length is 0 and request is async,
                        // we won't have a HttpWebResponse set on request object when this method is called
                        // http://referencesource.microsoft.com/#System/net/System/Net/HttpWebResponse.cs,525

                        // But we there will be CoreResponseData object that is either exception
                        // or the internal HTTP reponse representation having status, content and headers

                        var coreResponse = s_coreResponseAccessor(request);
                        if (coreResponse != null && s_coreResponseDataType.IsInstanceOfType(coreResponse))
                        {
                            HttpStatusCode status = s_coreStatusCodeAccessor(coreResponse);
                            WebHeaderCollection headers = s_coreHeadersAccessor(coreResponse);

                            // Manual creation of HttpWebResponse here is not possible as this method is eventually called from the
                            // HttpWebResponse ctor. So we will send Stop event with the Status and Headers payload
                            // to notify listeners about response;
                            // We use two different names for Stop events since one event with payload type that varies creates
                            // complications for efficient payload parsing and is not supported by DiagnosicSource helper
                            // libraries (e.g. Microsoft.Extensions.DiagnosticAdapter)

                            s_instance.RaiseResponseEvent(request, status, headers);
                        }
```

This response event does not check for the traceparent header however, to raise the event. like the `"System.Net.Http.Desktop.HttpRequestOut.Stop"` does (related issue: https://github.com/dotnet/runtime/issues/37124):

https://github.com/dotnet/runtime/blob/7565d60891e43415f5e81b59e50c52dba46ee0d7/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs#L668-L677

```c#
        private void RaiseResponseEvent(HttpWebRequest request, HttpStatusCode statusCode, WebHeaderCollection headers)
        {
            // Response event could be received several times for the same request in case it was redirected
            // IsLastResponse checks if response is the last one (no more redirects will happen)
            // based on response StatusCode and number or redirects done so far
            if (request.Headers.Get(RequestIdHeaderName) != null && IsLastResponse(request, statusCode))
            {
                this.Write(RequestStopExName, new { Request = request, StatusCode = statusCode, Headers = headers });
            }
        }
```

Still, we should at least see the start event and a span created for the call.

**Comment by russcam** [comment]
@gregkalapos I think it'd be good to merge this in, as it addresses the test hangs. There are still some intermittently flaky tests happening, but we can look at addressing those separately.

**Review by gregkalapos — APPROVED** [review]
Ok, let's merge this.

Do I see it correctly that the `HttpCallWithW3CActivityFormat` became green after you changed the logger? Super strange :)

**Comment by russcam** [comment]
> Do I see it correctly that the HttpCallWithW3CActivityFormat became green after you changed the logger? Super strange :)

It is strange. The successful test run output is as expected

```log
Elastic.Apm.Tests.HttpDiagnosticListenerTests.HttpCallWithW3CActivityFormat
    System.Diagnostics.DiagnosticSource version is System.Diagnostics.DiagnosticSource, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
    HttpCallWithW3CActivityFormat: {AbstractConfigurationReader (MockConfigSnapshot)} The agent was started without a service name. The service name will be automatically discovered.
    HttpCallWithW3CActivityFormat: {AbstractConfigurationReader (MockConfigSnapshot)} The agent was started without a service name. The automatically discovered service name is xunit_execution_desktop
    HttpCallWithW3CActivityFormat: {AbstractConfigurationReader (MockConfigSnapshot)} The agent was started without a service version. The service version will be automatically discovered.
    HttpCallWithW3CActivityFormat: {AbstractConfigurationReader (MockConfigSnapshot)} Failed to discover service version, the service version will be omitted.
    HttpCallWithW3CActivityFormat: {SystemInfoHelper} No /proc/self/cgroup found - the agent will not report container id
    HttpCallWithW3CActivityFormat: {HttpDiagnosticListener} Current runtime is detected as Full Framework. RuntimeInformation.FrameworkDescription: .NET Framework 4.8.4018.0
    HttpCallWithW3CActivityFormat: {DiagnosticInitializer} Subscribed Elastic.Apm.DiagnosticListeners.HttpDiagnosticListenerFullFrameworkImpl to `System.Net.Http.Desktop' events source
    HttpCallWithW3CActivityFormat: {AbstractConfigurationReader (MockConfigSnapshot)} No transaction sample rate provided. Defaulting to '1'
    HttpCallWithW3CActivityFormat: {Transaction} New Transaction instance created: Transaction{Id: 14f2cb02dff5c044, TraceId: a03a24070e70c44f92e5b7c613c5059a, ParentId: null, Name: Test, Type: Test, Outcome: Unknown, IsSampled: True}. IsSampled (True) is based on the given sampler (Sampler{ constant: True }). Start time: 2021-02-23 06:37:33.1505780 UTC (as timestamp: 1614062253150578)
    HttpCallWithW3CActivityFormat: {Tracer} Starting Transaction{Id: 14f2cb02dff5c044, TraceId: a03a24070e70c44f92e5b7c613c5059a, ParentId: null, Name: Test, Type: Test, Outcome: Unknown, IsSampled: True}
    Making request to: http://localhost:49215/
    HttpCallWithW3CActivityFormat: {HttpDiagnosticListenerImplBase} Called with key: `System.Net.Http.Desktop.HttpRequestOut.Start'
    HttpCallWithW3CActivityFormat: {AbstractConfigurationReader (MockConfigSnapshot)} Using default ServerUrl: http://localhost:8200/
    HttpCallWithW3CActivityFormat: {HttpDiagnosticListenerImplBase} Processing start event... Request URL: http://localhost:49215/
    HttpCallWithW3CActivityFormat: {AbstractConfigurationReader (MockConfigSnapshot)} No transaction max spans provided. Defaulting to '500'
    HttpCallWithW3CActivityFormat: {Span.c4ec8baa6cfdb847} New Span instance created: Span{Id: c4ec8baa6cfdb847, TransactionId: 14f2cb02dff5c044, ParentId: 14f2cb02dff5c044, TraceId: a03a24070e70c44f92e5b7c613c5059a, Name: GET localhost, Type: external, Outcome: Unknown, IsSampled: True}. Start time: 2021-02-23 06:37:33.1564680 UTC (as timestamp: 1614062253156468). Parent span: (null)
    HttpCallWithW3CActivityFormat: {Transaction} Starting Span{Id: c4ec8baa6cfdb847, TransactionId: 14f2cb02dff5c044, ParentId: 14f2cb02dff5c044, TraceId: a03a24070e70c44f92e5b7c613c5059a, Name: GET localhost, Type: external, Outcome: Unknown, IsSampled: True}
    HttpCallWithW3CActivityFormat: {AbstractConfigurationReader (MockConfigSnapshot)} UseElasticTraceparentHeader configuration option doesn't have a valid value - using default: True
    HttpCallWithW3CActivityFormat: {HttpDiagnosticListenerImplBase} Called with key: `System.Net.Http.Desktop.HttpRequestOut.Stop'
    HttpCallWithW3CActivityFormat: {HttpDiagnosticListenerImplBase} Processing stop event... Request URL: http://localhost:49215/
    HttpCallWithW3CActivityFormat: {Span.c4ec8baa6cfdb847} Ended Span{Id: c4ec8baa6cfdb847, TransactionId: 14f2cb02dff5c044, ParentId: 14f2cb02dff5c044, TraceId: a03a24070e70c44f92e5b7c613c5059a, Name: GET localhost, Type: external, Outcome: Success, IsSampled: True}. Start time: 2021-02-23 06:37:33.1564680 UTC (as timestamp: 1614062253156468), End time: 2021-02-23 06:37:33.1574480 UTC (as timestamp: 1614062253157448), Duration: 0.98ms
    HttpCallWithW3CActivityFormat: {Transaction} Ended Transaction{Id: 14f2cb02dff5c044, TraceId: a03a24070e70c44f92e5b7c613c5059a, ParentId: null, Name: Test, Type: Test, Outcome: Unknown, IsSampled: True}. Start time: 2021-02-23 06:37:33.1505780 UTC (as timestamp: 1614062253150578), End time: 2021-02-23 06:37:33.1828670 UTC (as timestamp: 1614062253182867), Duration: 32.289ms
```

**Pull request merged**